File: src/neetcode/Java/_50Pow_x_n_.java
package neetcode.Java;

public class _50Pow_x_n_ {
    public double myPow1(double x, int n) {
        return Math.pow(x, n);
    }

    public double myPow(double x, int n) {
        if (n == 0) return 1;
        if (n == 1) return x;
        if (n > 0) {
            return n % 2 == 0 ? myPow(x * x, n / 2) : x * myPow(x * x, n / 2);
        } else if (n == Integer.MIN_VALUE) {
            return myPow(x * x, n / 2);
        } else {
            return myPow(1.0 / x, -1 * n);
        }
    }

    public static void main(String[] args) {
        _50Pow_x_n_ obj = new _50Pow_x_n_();
        System.out.println(obj.myPow(2.0, -2147483648));
    }
}

File: src/neetcode/kotlin/_50Pow_x_n_.kt
package neetcode.kotlin

import kotlin.math.pow

class _50Pow_x_n_ {
    fun myPow1(x: Double, n: Int): Double {
        return x.pow(n.toDouble())
    }

    fun myPow(x: Double, n: Int): Double {
        if (n == 0) return 1.0
        if (n == 1) return x
        return if (n > 0) {
            if (n % 2 == 0) myPow(x * x, n / 2) else x * myPow(x * x, n / 2)
        } else if (n == Int.MIN_VALUE) {
            myPow(x * x, n / 2)
        } else {
            myPow(1.0 / x, -1 * n)
        }
    }

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            val obj = _50Pow_x_n_()
            println(obj.myPow(2.0, -2147483648))
        }
    }
}
